core/trino-main/src/main/java/io/trino/connector/CoordinatorDynamicCatalogManager.java

```diff
@@ -169,6 +180,44 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> catalo
         }
     }
 
+    @Override
+    public void pruneCatalogs(Set<CatalogHandle> catalogsInUse)
+    {
+        List<CatalogConnector> removedCatalogs = new ArrayList<>();
+        catalogsUpdateLock.lock();
+        try {
+            if (state == State.STOPPED) {
+                return;
+            }
+            for (Iterator<Entry<CatalogHandle, CatalogConnector>> iterator = allCatalogs.entrySet().iterator(); iterator.hasNext(); ) {
+                Entry<CatalogHandle, CatalogConnector> entry = iterator.next();
+
+                Catalog activeCatalog = activeCatalogs.get(entry.getKey().getCatalogName());
+                if (activeCatalog != null && activeCatalog.getCatalogHandle().equals(entry.getKey())) {
```

Comment:
What does it mean for `activeCatalogs` to contain the catalog, but for the handle not to match? It seems like we could remove from `allCatalogs` but not `activeCatalogs` in this case.